File: othello/othello.py
import numpy as np

OUTER = -1
EMPTY = 0
BLACK = 1
WHITE = 2

UP, DOWN, LEFT, RIGHT = -10, 10, -1, 1
UP_RIGHT, DOWN_RIGHT, DOWN_LEFT, UP_LEFT = -9, 11, 9, -11
DIRECTIONS = (UP, UP_RIGHT, RIGHT, DOWN_RIGHT, DOWN, DOWN_LEFT, LEFT, UP_LEFT)

class Othello:
    def __init__(self):
        
        self.__squares = np.array([i for i in range(11, 89) if 1 <= (i % 10) <= 8])
        self.__board, self.__empties = self.__new_board()
        self.__current_player = BLACK
        
    def player(self):
        return self.__current_player
    
    def opponent(self):
        return BLACK if self.__current_player is WHITE else WHITE
    
    def legal_moves(self):
        return [move for move in self.__empties if self.__has_bracket(move)]

    def play(self, move):
        self.__board[move] = self.__current_player
        self.__empties.remove(move)
        
        
        for direction in DIRECTIONS:
            self.__flip(move, direction)
            
        self.__current_player = self.opponent()
        
    def score(self):
        black_pieces = np.count_nonzero(self.__board == BLACK)
        white_pieces = np.count_nonzero(self.__board == WHITE)
        
        return black_pieces - white_pieces
            
    def __flip(self, move, direction):
        bracket = self.__find_bracket(move, direction)
        if not bracket:
            return
        square = move + direction
        while square != bracket:
            self.__board[square] = self.__current_player
            square += direction
    
    
    def __find_bracket(self, square, direction):
        bracket = square + direction
        if self.__board[bracket] == self.__current_player:
            return None

        while self.__board[bracket] == self.opponent():
            bracket += direction

        return None if self.__board[bracket] in (OUTER, EMPTY) else bracket
    
    def __has_bracket(self, move):
            return any(map(lambda d: self.__find_bracket(move, d), DIRECTIONS))
    
    
    def __new_board(self):
        board = -1 * np.ones(100)
        board[self.__squares] = EMPTY
        
        empties = set()
        
        for val in self.__squares:
            empties.add(val)

        board[44], board[45] = WHITE, BLACK
        board[54], board[55] = BLACK, WHITE
        
        empties.discard(44)
        empties.discard(45)
        empties.discard(54)
        empties.discard(55)
        
        return board, empties
    
    def __str__(self):
        
        board = [OUTER] * 100
        for i in range(100):
            if self.__board[i] == OUTER:
                board[i] = '?'
            elif self.__board[i] == EMPTY:
                board[i] = '.'
            elif self.__board[i] == BLACK:
                board[i] = 'o'
            else:
                board[i] = 'X'
        
        rep = ''
        rep += '  %s\n' % ' '.join(map(str, range(1, 9)))
        for row in range(1, 9):
            begin, end = 10*row + 1, 10*row + 9
            rep += '%d %s\n' % (row, ' '.join(board[begin:end]))
        return rep

File: fast-mcts/Othello.cpp
#include "Othello.hpp"

Othello::Othello() {}

Othello::~Othello() {}

void Othello::play(int8_t move) {

    this->_board[move] = this->_player;
    this->_empties[move] = 0;

    for(int8_t i = 0; i < 8; i++) {
        this->flip(move, _OTHELLO_DIRECTIONS[i]);
    }

    this->_player *= -1;
}

std::vector<int8_t> Othello::moves() const {
    std::vector<int8_t> moves;

    for(int8_t i = 0; i < 100; i++) {
        if(this->_empties[i] == 1 && this->hasBracket(i)) {
            moves.push_back(i);
        }
    }

    return moves;
}

// Black wins on score > 0, White wins on score < 0, Draw on score == 0
int8_t Othello::score() {
    int8_t diff = 0;

    for(int8_t i = 0; i < 100; i++) {
        if(this->_board[i] == BLACK) diff++;
        else if (this->_board[i] == WHITE) diff--;
    }

    return diff;
}

torch::Tensor Othello::state() const {

    torch::Tensor gameState = torch::zeros({3, 8, 8});
    int8_t array_index;

    for(int8_t i = 0; i < 8; i++) {
        for(int8_t j = 0; j < 8; j++) {
            array_index = ((i + 1) * 10) + j + 1;

            if (this->_board[array_index] == BLACK) gameState[0][i][j] = 1;
            else if (this->_board[array_index] == WHITE) gameState[1][i][j] = 1;
        }
    }

    if (this->_player == BLACK) gameState[2] = 1;

    return gameState;
}

void Othello::render() const {
    

    for (int8_t i = 0; i < 8; i++) {
        for(int8_t j = 0; j < 8; j++) {
            int8_t array_index = ((i + 1) * 10) + j + 1;
            if (this->_board[array_index] == BLACK) {
                std::cout << "o ";
            } else if (this->_board[array_index] == WHITE) {
                std::cout << "X ";
            } else if (this->_board[array_index] == EMPTY) {
                std::cout << ". ";
            }
        }
        std::cout << std::endl;
    }
}

Game* Othello::copy() {
    Othello* newGame = new Othello();

    newGame->setBoard(this->_board);
    newGame->setEmpties(this->_empties);
    newGame->setPlayer(this->_player);

    return (Game*) newGame;
}

int8_t Othello::player() const {
    return this->_player;
}

//TODO: double check if this is working 
int8_t Othello::moveToAction(int8_t move) const {
    return move - 11 - (2 * ((int8_t)move/10 - 1));
}

//TODO: double check if this is working 
int8_t Othello::actionToMove(int8_t action) const {
    return (10 *(int8_t)floor(action / 8)) + 11 + action % 8;
}

/* PRIVATE STUFF */
void Othello::flip(int8_t move, int8_t direction) {
    int8_t bracket = this->findBracket(move, direction);

    if(bracket == -1) return;

    int8_t square = move + direction;

    while (square != bracket) {
        this->_board[square] = this->_player;
        square += direction;
    }
}

int8_t Othello::findBracket(int8_t square, int8_t direction) const {
    int8_t bracket = square + direction;

    if (this->_board[bracket] == this->_player) return -1;

    int8_t op = -1 * this->_player;

    while (this->_board[bracket] == op) bracket += direction;

    if (this->_board[bracket] == OUTER || this->_board[bracket] == EMPTY) return -1;

    return bracket;
}

bool Othello::hasBracket(int8_t move) const {
    for(int8_t i = 0; i < 8; i++) {
        if (this->findBracket(move, _OTHELLO_DIRECTIONS[i]) != -1) return true;
    }
    return false;
}

void Othello::setBoard(int8_t *newBoard) {
    memcpy(this->_board, newBoard, 100);
}

void Othello::setPlayer(int8_t newPlayer) {
    this->_player = newPlayer;
}

void Othello::setEmpties(int8_t *newEmpties) {
    memcpy(this->_empties, newEmpties, 100);
}
